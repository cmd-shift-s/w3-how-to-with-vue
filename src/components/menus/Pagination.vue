<!-- http://fareez.info/blog/pagination-component-using-vuejs/ -->
<template lang="html">
  <div class="pagination">
    <a href="javascript:void(0)" @click.prevent="pageChanged(1)">&laquo;</a>
    <a href="javascript:void(0)" v-for="n in paginationRange" :class="activePage(n)" @click.prevent="pageChanged(n)" :key="n">{{n}}</a>
    <a href="javascript:void(0)" @click.prevent="pageChanged(lastPage)">&raquo;</a>
  </div>
</template>

<script>
export default {
  name: 'pagination',
  props: {
    // Current Page
    currentPage: {
      type: Number,
      required: true
    },

    // Total number of pages
    totalPages: Number,

    // Items per page
    itemsPerPage: Number,

    // Total items
    totalItems: Number,

    // Visible Pages
    visiblePages: {
      type: Number,
      default: 5
    }
  },
  computed: {
    lastPage() {
      if (this.totalPages) {
        return this.totalPages
      } else {
        return this.totalItems % this.itemsPerPage === 0
          ? this.totalItems / this.itemsPerPage
          : Math.floor(this.totalItems / this.itemsPerPage) + 1
      }
    },
    paginationRange() {
      const page = this.lastPage - this.visiblePages + 1
      const start =
        this.currentPage - this.visiblePages / 2 <= 0
          ? 1
          : this.currentPage + this.visiblePages / 2 > this.lastPage
            ? page >= 1 ? page : 1
            : Math.ceil(this.currentPage - this.visiblePages / 2)

      const range = []

      for (let i = 0; i < this.visiblePages && i < this.lastPage; i++) {
        range.push(start + i)
      }

      return range
    }
  },
  methods: {
    activePage(pageNum) {
      return this.currentPage === pageNum ? 'is-active' : ''
    },
    pageChanged(pageNum) {
      this.$emit('page-changed', pageNum)
    }
  }
}
</script>

<style lang="scss">
.pagination {
  /* Pagination links */
  a {
    color: black;
    float: left;
    padding: 8px 16px;
    text-decoration: none;
    transition: background-color .3s;

    /* Style the active/current link */
    &.is-active {
      background-color: #4CAF50;
      color: white;
    }

    /* Add a grey background color on mouse-over */
    &:hover:not(.is-active) {
      background-color: #ddd;
    }
  }
}
</style>
