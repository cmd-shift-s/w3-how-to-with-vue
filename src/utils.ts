import MarkdownIt from 'markdown-it'
import hljs from 'highlight.js'
import 'highlight.js/styles/vs.css'

export const md = new MarkdownIt({
  highlight (code, lang) {
    return hljs.highlight(lang, code).value
  }
})
