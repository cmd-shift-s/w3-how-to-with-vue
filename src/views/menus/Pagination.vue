<template lang="html">
  <page
    title="Pagination"
    intro="a pagination with CSS">

    <template slot="header">
    <h2>How To Create a Pagination</h2>
    </template>

    <sample>
      <pagination :current-page="curPage" :total-pages="6" @page-changed="pageChanged" :visiblePages="10"></pagination>
    </sample>

    <template slot="examples">
    <hr>
    <step no="1" html></step>
    <example-code :code="examples[0]"></example-code>
    <step no="2" css></step>
    <example-code :code="examples[1]"></example-code>
    </template>
  </page>
</template>

<script>
import Pagination from '@/components/menus/Pagination.vue'

export default {
  name: 'pagination-page',
  data() {
    return {
      curPage: 2,
      examples: [
        [
          '``` html',
          '<div class="pagination">',
          '  <a href="#">&laquo;</a>',
          '  <a href="#">1</a>',
          '  <a class="active" href="#">2</a>',
          '  <a href="#">3</a>',
          '  <a href="#">4</a>',
          '  <a href="#">5</a>',
          '  <a href="#">6</a>',
          '  <a href="#">&raquo;</a>',
          '</div>',
          '```'
        ],
        [
          '``` css',
          '/* Pagination links */',
          '.pagination a {',
          '  color: black;',
          '  float: left;',
          '  padding: 8px 16px;',
          '  text-decoration: none;',
          '  transition: background-color .3s;',
          '}',
          '',
          '/* Style the active/current link */',
          '.pagination a.active {',
          '  background-color: #4CAF50;',
          '  color: white;',
          '}',
          '',
          '/* Add a grey background color on mouse-over */',
          '.pagination a:hover:not(.active) {background-color: #ddd;}',
          '```'
        ]
      ]
    }
  },
  methods: {
    pageChanged(page) {
      this.curPage = page
    }
  },
  components: {
    Pagination
  }
}
</script>

<style lang="css">
</style>
