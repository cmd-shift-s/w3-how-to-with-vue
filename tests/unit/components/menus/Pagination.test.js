import { mount, renderer } from '@/tests/unit/utils'
import Pagination from '@/components/menus/Pagination.vue'

describe('Pagination.vue', () => {
  it('render default', () => {
    const wrap = mount(Pagination, {
      propsData: {
        currentPage: 1
      }
    })

    expect(wrap.classes()).toContain('pagination')

    renderer.renderToString(wrap.vm, (err, str) => {
      if (err) console.error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('props#total-pages', () => {
    const totalPages = 5
    const wrap = mount(Pagination, {
      propsData: {
        currentPage: 1,
        totalPages
      }
    })

    expect(wrap.findAll('a').length).toBe(totalPages + 2) // 2: <<, >>

    renderer.renderToString(wrap.vm, (err, str) => {
      if (err) console.error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('emit$page-changed', () => {
    const wrap = mount(Pagination, {
      propsData: {
        currentPage: 1,
        totalPages: 5
      }
    })

    wrap.findAll('a').at(3).trigger('click')

    const pageChanged = wrap.emitted()['page-changed']
    expect(pageChanged).toBeTruthy()
    expect(pageChanged.length).toBe(1)
    expect(pageChanged[0]).toEqual([3])
  })
})
