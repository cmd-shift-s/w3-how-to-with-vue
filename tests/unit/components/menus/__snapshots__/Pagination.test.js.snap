// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination.vue props#total-pages 1`] = `<div class="pagination"><a href="javascript:void(0)">«</a> <a href="javascript:void(0)" class="is-active">1</a><a href="javascript:void(0)">2</a><a href="javascript:void(0)">3</a><a href="javascript:void(0)">4</a><a href="javascript:void(0)">5</a> <a href="javascript:void(0)">»</a></div>`;

exports[`Pagination.vue render default 1`] = `<div class="pagination"><a href="javascript:void(0)">«</a> <a href="javascript:void(0)">»</a></div>`;
